### PR TITLE
feat(backend): Migrate to vitest

### PR DESCRIPTION
**Pull request by jacekradko**
## Description

Spiking out `vitest` test runner as a replacement for `qunit` in `@clerk/backend`.

Changes
- Using `vitest` runner
- Using `msw` instead of mocking `fetch`

Related: https://linear.app/clerk/issue/SDKI-694/refactor-clerkbackend-to-use-vitest-instead-of-qunit

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/backend/src/api/request.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ export function buildRequest(options: BuildRequestOptions) {
     let res: Response | undefined;
     try {
       if (formData) {
-        res = await runtime.fetch(finalUrl.href, {
+        res = await (process.env.NODE_ENV === 'test' ? fetch : runtime.fetch)(finalUrl.href, {
```
Comment (jacekradko) [inline packages/backend/src/api/request.ts:91]: Will need to think about a more ergonomic way of accomplishing this. With the way `fetch` is handled in the package `msw` can't intercept requests without it for now

File: packages/backend/.eslintrc.js [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
     'import/no-unresolved': ['error', { ignore: ['^#'] }],
   },
   overrides: [
     {
       files: ['*.test.ts'],
       rules: {
-        // TODO: It's an issue specific to QUnit tests
-        '@typescript-eslint/unbound-method': 'off',
+        // // TODO: It's an issue specific to QUnit tests
+        // '@typescript-eslint/unbound-method': 'off',
```
Comment (jacekradko) [inline packages/backend/.eslintrc.js:12]:
```suggestion
```

File: .vscode/launch.json [hunk 3]
```diff
@@ -21,6 +21,19 @@
...
+        "type": "node",
+        "request": "launch",
+        "program": "${workspaceFolder}/scripts/vitest-debug.mjs",
+        "args": ["${file}"],
+        "cwd": "${workspaceFolder}",
+        "console": "integratedTerminal",
+        "internalConsoleOptions": "neverOpen",
+        "autoAttachChildProcesses": true,
+        "smartStep": true,
+        "skipFiles": ["<node_internals>/**", "**/node_modules/**"]
```
Comment (jacekradko) [inline .vscode/launch.json:36]: Added a debug configuration for any `vitest` spec

File: packages/backend/src/api/request.ts (hunk 1 above)
Comment (jacekradko) [reply packages/backend/src/api/request.ts:89]: @BRKalow Any thoughts on how you would like to move forward here?

File: packages/backend/src/tokens/__tests__/keys.test.ts [hunk 4]
```diff
@@ -11,212 +11,193 @@ import {
...
     });
+    expect(jwk).toMatchObject(mockRsaJwk);
+    jwk = await loadClerkJWKFromRemote({
+      secretKey: 'deadbeef',
+      kid: mockRsaJwkKid,
+    });
+    expect(jwk).toMatchObject(mockRsaJwk);
+  });
+
+  it.skip('retries five times with exponential back-off policy to fetch JWKS before it fails', async () => {
```
Comment (jacekradko) [inline packages/backend/src/tokens/__tests__/keys.test.ts:101]: Need to loop back to this test

File: packages/backend/src/tokens/__tests__/request.test.ts [hunk 5]
```diff
@@ -25,850 +25,1008 @@ import type { AuthenticateRequestOptions, OrganizationSyncOptions } from '../typ
...
+        pass: true,
+      };
+    } else {
+      return {
+        message: () => `expected to be signed in`,
+        pass: false,
+      };
+    }
+  },
+  toBeSignedInToAuth(received) {
```
Comment (jacekradko) [inline packages/backend/src/tokens/__tests__/request.test.ts:172]: This matcher does not seem to be working correctly. 👀

File: packages/backend/src/api/request.ts (hunk 1 above)
Comment (BRKalow) [reply packages/backend/src/api/request.ts:89]: @jacekradko can we try to consolidate into the runtime module? so at least usage of runtime.fetch doesn't need to be aware of the `NODE_ENV`

File: packages/backend/vitest.setup.mts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { beforeAll, afterEach, afterAll } from 'vitest';
+import { server } from './src/mock-server';
+
+beforeAll(() => server.listen());
+afterEach(() => server.resetHandlers());
+afterAll(() => server.close());
+
+server.events.on('request:start', ({ request }) => {
+  // FOR DEBUGGING
+  // console.log('MSW intercepted:', request.method, request.url);
```
Comment (BRKalow) [inline packages/backend/vitest.setup.mts:10]: 👀 want to leave this? if yes, maybe we put it behind an env var or something?

File: packages/backend/src/api/__tests__/factory.test.ts [hunk 7]
```diff
@@ -1,257 +1,184 @@
...
+    server.use(
+      http.get(`https://api.clerk.test/v1/users`, () => {
+        return HttpResponse.json([userJson]);
+      }),
+    );
+    server.use(
+      http.get(`https://api.clerk.test/v1/users/count`, () => {
+        return HttpResponse.json({ object: 'total_count', total_count: 2 });
+      }),
+    );
```
Comment (BRKalow) [inline packages/backend/src/api/__tests__/factory.test.ts:41]: much nicer mocking solution ✨

File: packages/backend/vitest.setup.mts (hunk 6 above)
Comment (jacekradko) [reply packages/backend/vitest.setup.mts:10]: @BRKalow I am fine either way. My thought was leaving it commented out if devs wanted to debug this in the future, but an ENV might be better. Do we have a naming standard for ENV vars?

File: packages/backend/src/jwt/__tests__/signJwt.test.ts [hunk 8]
```diff
@@ -9,43 +9,40 @@ import {
...
     });
+    expect(data).toBeTruthy();
 
-    test('signs a JWT with a JWK formatted secret', async assert => {
-      const { data } = await signJwt(payload, signingJwks, {
-        algorithm: mockJwtHeader.alg,
-        header: mockJwtHeader,
-      });
-      assertOk(assert, data);
+    // @ts-ignore - data will be there
```
Comment (BRKalow) [inline packages/backend/src/jwt/__tests__/signJwt.test.ts:32]: 👀 why did we need to add this `@ts-ignore`?

File: packages/backend/src/jwt/__tests__/verifyJwt.test.ts [hunk 9]
```diff
@@ -12,123 +10,120 @@ import {
...
+const invalidTokenError = {
+  reason: 'token-invalid',
+  message: 'Invalid JWT form. A JWT consists of three parts separated by dots.',
+};
+
+describe('hasValidSignature(jwt, key)', () => {
+  it('verifies the signature with a JWK formatted key', async () => {
+    const { data: decodedResult } = decodeJwt(signedJwt);
+    // @ts-expect-error
+    const { data: signatureResult } = await hasValidSignature(decodedResult, publicJwks);
```
Comment (BRKalow) [inline packages/backend/src/jwt/__tests__/verifyJwt.test.ts:24]: 👀 curious why we need these `@ts-expect-error`s now where we didn't previously

File: packages/backend/vitest.setup.mts (hunk 6 above)
Comment (BRKalow) [reply packages/backend/vitest.setup.mts:10]: We do not. `DEBUG_MOCK_REQUESTS` maybe? And add to the README 👀

**Review by BRKalow — COMMENTED** [review]
Really nice work! Would be great to shift the `NODE_ENV` fetch check into the runtime module. Left a few minor questions.

File: packages/backend/src/__tests__/exports.test.ts [hunk 10]
```diff
@@ -1,56 +1,56 @@
...
+  it('should not include a breaking change', () => {
+    const exportedApiKeys = [
+      'SignJWTError',
+      'TokenVerificationError',
+      'TokenVerificationErrorAction',
+      'TokenVerificationErrorCode',
+      'TokenVerificationErrorReason',
+    ];
+
+    expect(Object.keys(errorExports).sort()).toEqual(exportedApiKeys);
```
[the remark this answers is not in the document]
Comment (jacekradko) [reply packages/backend/src/__tests__/exports.test.ts:26]: Agreed

File: packages/backend/vitest.setup.mts (hunk 6 above)
Comment (jacekradko) [reply packages/backend/vitest.setup.mts]: Addressed in [`ae16e84` (#4296)](https://github.com/clerk/javascript/pull/4296/commits/ae16e84a53e7109bf7772fb25d484dba09e30f73)

File: packages/backend/package.json [hunk 11]
```diff
@@ -84,14 +84,13 @@
...
     "lint:attw": "attw --pack . --ignore-rules false-cjs",
-    "test": "run-s clean:tests build:runtime build:tests test:node test:edge-runtime test:cloudflare-miniflare",
-    "test:node": "./tests/node/run.sh",
-    "test:edge-runtime": "./tests/edge-runtime/run.sh",
-    "test:cloudflare-miniflare": "tests/cloudflare-miniflare/run.sh",
+    "test": "run-s test:node test:edge-runtime test:cloudflare-miniflare",
+    "test:node": "vitest --environment node",
+    "test:edge-runtime": "vitest --environment edge-runtime",
+    "test:cloudflare-miniflare": "vitest --environment miniflare",
     "test:cloudflare-workerd": "tests/cloudflare-workerd/run.sh"
```
Comment (jacekradko) [inline packages/backend/package.json:94]:
```suggestion
```

File: packages/backend/src/api/__tests__/factory.test.ts (hunk 7 above)
Comment (jacekradko) [reply packages/backend/src/api/__tests__/factory.test.ts:41]: 😄

File: packages/backend/src/jwt/__tests__/verifyJwt.test.ts (hunk 9 above)
Comment (jacekradko) [reply packages/backend/src/jwt/__tests__/verifyJwt.test.ts]: Fixed

File: packages/backend/src/tokens/__tests__/keys.test.ts (hunk 4 above)
Comment (jacekradko) [reply packages/backend/src/tokens/__tests__/keys.test.ts]: Fixed!

**Review by BRKalow — APPROVED** [review]
:shipit:

File: packages/backend/src/api/request.ts [hunk 12]
```diff
@@ -88,7 +86,7 @@ export function buildRequest(options: BuildRequestOptions) {
...
```
[the remark this answers is not in the document]
Comment (jacekradko) [reply packages/backend/src/api/request.ts:90]: Done

File: packages/backend/src/api/request.ts [hunk 13]
```diff
@@ -100,7 +98,7 @@ export function buildRequest(options: BuildRequestOptions) {
         const hasBody = method !== 'GET' && bodyParams && Object.keys(bodyParams).length > 0;
         const body = hasBody ? { body: JSON.stringify(snakecaseKeys(bodyParams, { deep: false })) } : null;
 
-        res = await runtime.fetch(finalUrl.href, {
+        res = await (process.env.NODE_ENV === 'test' ? fetch : runtime.fetch)(finalUrl.href, {
```
[the remark this answers is not in the document]
Comment (jacekradko) [reply packages/backend/src/api/request.ts:103]: Done

**Pull request merged**